### PR TITLE
Add missing IArray operations

### PR DESCRIPTION
**Pull request by nicolasstucki**
* Add missing operations
* Adapt some existing operations
* Remove some `ClassTag` from existing operations that did not need them
* Added `IArrayOps.WithFilter` and `IArrayBuilder`

**TODO**
- [x] Solve #11358 and add missing overloads that receive `IArray`s
- [x] Fix double definition of `startsWith` (removed default parameters)
- [x] Should operations with implicit conversion keep those conversions? See `flatMap`, `WithFilter.flatMap`, `flatten`, `unzip`, `unzip3`, `transpose`.
- [x] Where should we place the `WithFilter` and `IArrayBuilder`?

**Left for later**
* Refactor collective extensions and add documentation

### CODE REVIEW

File: library/src/scala/IArray.scala [hunk 1]
```diff
@@ -102,7 +100,7 @@ object opaques:
 
   /** Flattens a two-dimensional array by concatenating all its rows
     * into a single array. */
-  extension [T](arr: IArray[T]) def flatten[U: ClassTag](using T => Iterable[U]): IArray[U] =
+  extension [T](arr: IArray[T]) def flatten[U: ClassTag](using asIterable: T => Iterable[U]): IArray[U] =
```
Comment (nicolasstucki) [inline library/src/scala/IArray.scala:103]: If we are against adding new implicit conversions should we simplify the definitions of `flatMap`, `WithFilter.flatMap`, `flatten`, `unzip`, `unzip3`, `transpose`?

For this one the alternative is to define it as
```scala
 extension [T](arr: IArray[Iterable[T]]) def flatten: IArray[T] = ...
```
We could also add an overload for 
```scala
 extension [T](arr: IArray[IArray[T]]) def flatten: IArray[T] = ...
```
@odersky @sjrd WDYT?

File: library/src/scala/IArray.scala [hunk 2]
```diff
@@ -535,4 +648,67 @@ object IArray {
     // The double type ascription is currently needed,
     // for some reason (see: https://scastie.scala-lang.org/sSsmOhKxSKym405MgNRKqQ)
     Array.unapplySeq((x: Sup[T]): Array[_ <: T])
-}
+
+  // TODO where should this class be defined? The WithFilter for Array is in ArrayOps.
+  /** A lazy filtered array. No filtering is applied until one of `foreach`, `map` or `flatMap` is called. */
+  class WithFilter[T](p: T => Boolean, xs: IArray[T]):
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: Now I also have `IArrayBuilder`

[the remark this answers is not in the document]
Comment (sjrd) [reply library/src/scala/IArray.scala]: The `WithFilter` of `Array` is in `ArrayOps` because that's where the `def withFilter` method is. In this case, since I guess `withFilter` is in `IArray`, there's no reason to put `WithFilter` anywhere else.

File: library/src/scala/IArray.scala (hunk 1 above)
Comment (sjrd) [reply library/src/scala/IArray.scala]: We'll need an implicit conversion from `IArray[T]` to `Iterable[T]` anyway, because I can to be able to give `IArray`s in the `flatMap` methods of `List` and other collections.

File: library/src/scala/IArray.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: Then I guess we dot the same for `IArrayBuilder`

File: library/src/scala/IArray.scala (hunk 1 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: Should the conversion be to `Iterable[T]` or to `immutable.ArraySeq[T]` directly?

Comment (sjrd) [reply library/src/scala/IArray.scala]: `immutable.ArraySeq` sounds reasonable to me.

Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: This allowed me to remove many redundant operations. See the second commit.

File: library/src/scala/IArray.scala [hunk 3]
```diff
@@ -135,9 +131,6 @@ object IArray:
 
   /** Finds index of first occurrence of some value in this array after or at some start index. */
   extension [T](arr: IArray[T]) def indexOf(elem: T, from: Int = 0): Int =
-    // `asInstanceOf` needed because `elem` does not have type `arr.T`
-    // We could use `arr.iterator.indexOf(elem, from)` or `arr.indexWhere(_ == elem, from)`
-    // but these would incur some overhead.
     genericArrayOps(arr).indexOf(elem.asInstanceOf, from)
```
Comment (sjrd) [inline library/src/scala/IArray.scala:134]: The comment disappeared, but the `asInstanceOf` remains. Probably the comment should be preserved as well.

File: library/src/scala/IArray.scala [hunk 4]
```diff
@@ -170,7 +163,6 @@ object IArray:
 
   /** Finds index of last occurrence of some value in this array before or at a given end index. */
   extension [T](arr: IArray[T]) def lastIndexOf(elem: T, end: Int = arr.length - 1): Int =
-    // see: same issue in `indexOf`
     genericArrayOps(arr).lastIndexOf(elem.asInstanceOf, end)
```
Comment (sjrd) [inline library/src/scala/IArray.scala:166]: Same comment as in `indexOf`.

File: library/src/scala/IArray.scala [hunk 5]
```diff
@@ -260,59 +248,125 @@ object IArray:
...
+    def zip[U](that: IArray[U]): IArray[(T, U)] = genericArrayOps(arr).zip(that)
+    def zip[U](that: IterableOnce[U]): IArray[(T, U)] = genericArrayOps(arr).zip(that)
+    def zipAll[T1 >: T, U](that: IArray[U], thisElem: T1, thatElem: U): IArray[(T1, U)] = genericArrayOps(arr).zipAll(that, thisElem, thatElem)
+    def zipAll[T1 >: T, U](that: Iterable[U], thisElem: T1, thatElem: U): IArray[(T1, U)] = genericArrayOps(arr).zipAll(that, thisElem, thatElem)
+    def zipWithIndex: IArray[(T, Int)] = genericArrayOps(arr).zipWithIndex
+  end extension
+
+  extension [T](arr: IArray[T])
+    def transpose[U](implicit asArray: T => IArray[U]): IArray[IArray[U]] =
+      genericArrayOps(arr).transpose(using asArray.asInstanceOf[T => Array[U]])
```
Comment (sjrd) [inline library/src/scala/IArray.scala:316]: Any reason this one is not in the big collective extension above?

File: library/src/scala/IArray.scala [hunk 6]
```diff
@@ -260,59 +248,125 @@ object IArray:
...
+  extension [T, U >: T: ClassTag](prefix: IArray[T])
+    def ++:(arr: IArray[U]): IArray[U] = genericArrayOps(arr).prependedAll(prefix)
+
+  extension [T, U >: T: ClassTag](x: T)
+    def +:(arr: IArray[U]): IArray[U] = genericArrayOps(arr).prepended(x)
 
   /** Conversion from IArray to immutable.ArraySeq */
-  extension [T](arr: IArray[T]) def toSeq: immutable.ArraySeq[T] =
-    immutable.ArraySeq.unsafeWrapArray(arr.asInstanceOf[Array[T]])
+  given genericWrapIArray[T](using DummyImplicit): Conversion[IArray[T], scala.collection.immutable.ArraySeq[T]] =
```
Comment (sjrd) [inline library/src/scala/IArray.scala:328]: Why the `DummyImplicit`s, here and below?

File: library/src/scala/IArray.scala [hunk 7]
```diff
@@ -363,16 +417,32 @@ object IArray:
...
    *
    *  @param xss the given immutable arrays
    *  @return   the array created from concatenating `xss`
    */
   def concat[T: ClassTag](xss: IArray[T]*): IArray[T] =
-    // `Array.concat` should arguably take in a `Seq[Array[_ <: T]]`,
-    // but since it currently takes a `Seq[Array[T]]` we have to perform a cast,
-    // knowing tacitly that `concat` is not going to do the wrong thing.
-    Array.concat[T](xss.asInstanceOf[Seq[Array[T]]]: _*)
+    Array.concat[T](xss.asInstanceOf[immutable.Seq[Array[T]]]: _*)
```
Comment (sjrd) [inline library/src/scala/IArray.scala:445]: Same here: the comment should probably stay. Also, why the change `Seq` -> `immutable.Seq` (they're the same)?

File: library/src/scala/IArray.scala [hunk 8]
```diff
@@ -521,4 +601,575 @@ object IArray:
...
+    def withFilter(q: T => Boolean): WithFilter[T] = new WithFilter[T](a => p(a) && q(a), xs)
+
+  end WithFilter
+
+  /** A builder class for immutable arrays.
+  *
+  *  @tparam T    the type of the elements for the builder.
+  */
+  @SerialVersionUID(3L)
+  sealed abstract class IArrayBuilder[T]
```
Comment (sjrd) [inline library/src/scala/IArray.scala:670]: Hum, now that I see it here, I don't think it's the right place for this one, in fact. `ArrayBuilder` is a top-level class in `scala.collection.mutable`, even though `Array` is in `scala.*`. By analogy, `IArrayBuilder` should also be a top-level class in `scala.collection.mutable` (yes, `mutable`, like `StringBuilder`).

File: library/src/scala/IArray.scala (hunk 1 above)
Comment (odersky) [reply library/src/scala/IArray.scala]: The problem with using `Conversion` instances is that you get a feature warning. 
```
there were 1 feature warning(s); re-run with -feature for details
```
If you run with `-feature` you get something like this
```
  |Use of implicit conversion given instance genericWrapAnyRefIArray in object IArray should be enabled
  |by adding the import clause 'import scala.language.implicitConversions'
  |or by setting the compiler option -language:implicitConversions.
  |See the Scala docs for value scala.language.implicitConversions for a discussion
  |why the feature should be explicitly enabled.
```
To prevent this you can use an old-style implicit def. Or, alternatively, change `checkImplicitConversionUseOK` to make a special case for conversions coming from `IArray`. But the second alternative requires a discussion what the policy wrt implicit conversions should be. So I think it's easiest for now to simply revert to implicit defs for the conversions here.

**Review by sjrd — CHANGES_REQUESTED** [review]

File: library/src/scala/IArray.scala [hunk 9]
```diff
@@ -521,4 +601,575 @@ object IArray:
...
+        }
+        size += k
+      } else if (k < 0) super.addAll(xs)
+      this
+    }
+  }
+
+  /** A companion object for array builders.
+  */
+  object IArrayBuilder {
```
Comment (odersky) [inline library/src/scala/IArray.scala:725]: There should be a way to re-use array builders here. I believe there is an operation where you create a builder by using an existing builder and adding a post-processing step. In this case the post-processing step would be simply `asInstanceOf[IArray]`. Can we use that one?

**Review by odersky — CHANGES_REQUESTED** [review]

File: library/src/scala/IArray.scala (hunk 9 above)
Comment (odersky) [reply library/src/scala/IArray.scala:725]: We couls still have specialized versions for each primitive type, but we would not need to manually re-implement all the methods.

File: library/src/scala/IArray.scala (hunk 5 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala:316]: No.

File: library/src/scala/IArray.scala (hunk 4 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala:166]: Added back

File: library/src/scala/IArray.scala (hunk 3 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala:134]: Added back

File: library/src/scala/IArray.scala (hunk 7 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala:445]: I added the import to `import scala.collection._` which shadowed the default one.

File: library/src/scala/IArray.scala (hunk 6 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala:328]: If I don't I get an error at call site

```scala
[error] -- [E007] Type Mismatch Error: library/src/scala/runtime/TupleXXL.scala:26:18 
[error] 26 |        while i < es.length do
[error]    |                  ^^
[error]    |Found:    (TupleXXL.this.es : IArray[Object])
[error]    |Required: ?{ length: ? }
[error]    |Note that implicit extension methods cannot be applied because they are ambiguous;
[error]    |both object IArray in package scala and given instance genericWrapIArray in object IArray provide an extension method `length` on (TupleXXL.this.es : IArray[Object])
```

Maybe they should be placed in Predef as are the ones for Arrays. But adding them to `stdLibPatches/Predef.scala` crashed the compiler.

File: library/src/scala/IArray.scala (hunk 9 above)
Comment (sjrd) [reply library/src/scala/IArray.scala:725]: Indeed! This should probably be as easy as
```scala
@inline def makeIArrayBuilder[T: ClassTag]: Builder[T, IArray[T]] =
  mutable.ArrayBuilder.make[T].mapResult(r => r: IArray[T])
```

File: library/src/scala/IArray.scala (hunk 8 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: Moved to `scala.collection.mutable`

File: library/src/scala/IArray.scala (hunk 1 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: You mean `flatMap` as in the following example?
```scala
for
  x <- List(1, 2)
  y <- IArray(1, 2)
yield ...
```
For this, we do not need the implicit conversion.

File: library/src/scala/IArray.scala (hunk 9 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: I reimplemeted `IArrayBuilder` and moved it to `scala.collection.mutable`.

File: library/src/scala/IArray.scala (hunk 7 above)
Comment (sjrd) [reply library/src/scala/IArray.scala]: That is usually a dangerous import, which is typically not recommended. It means that you can get the non-immutable types by default in all sorts of places without meaning too. Consider either important exactly what you need, like
```scala
import scala.collection.Iterable
```
or just use `collection.Iterable` when you need the non-immutable ones.

File: library/src/scala/collection/mutable/IArrayBuilder.scala [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+
+/** A companion object for array builders. */
+object IArrayBuilder {
+
+  /** Creates a new arraybuilder of type `T`.
+  *
+  *  @tparam T     type of the elements for the array builder, with a `ClassTag` context bound.
+  *  @return       a new empty array builder.
+  */
+  @inline def make[T: ClassTag]: IArrayBuilder[T] = {
```
Comment (sjrd) [inline library/src/scala/collection/mutable/IArrayBuilder.scala:42]: I take it that using `mapResult` didn't work?

**Review by sjrd — APPROVED** [review]
LGTM other than the dangerous import, and possibly using `mapResult` for the builder.

File: library/src/scala/IArray.scala (hunk 1 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: I used `implicit def`s and aligned the implementation with the one for Arrays

File: library/src/scala/IArray.scala (hunk 6 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: Avoided this with `implicit def`

File: library/src/scala/IArray.scala (hunk 7 above)
Comment (nicolasstucki) [reply library/src/scala/IArray.scala]: Will do

File: library/src/scala/collection/mutable/IArrayBuilder.scala (hunk 10 above)
Comment (nicolasstucki) [reply library/src/scala/collection/mutable/IArrayBuilder.scala:42]: It works, but the signature must be `Builder[T, IArray[T]]` and it adds an extra indirection. 

This would mean that we do not have the efficient `addAll(xs: IArray[T])` and `(xs: IArray[T], offset: Int, length: Int)` on that builder.

Comment (nicolasstucki) [reply library/src/scala/collection/mutable/IArrayBuilder.scala:42]: I guess we could also have `addAll(xs: Array[T])`

Comment (nicolasstucki) [reply library/src/scala/collection/mutable/IArrayBuilder.scala:42]: We have 2 options:
* Return `IArrayBuilder[T]`
  * Lareger code
  * Efficient `add` for `IArray`/`Array`
  * Possible to create a specialized `IArrayBuilder[T]` for primitive types
* Return `Builder[T, IArray[T]]`
  * Trivial code
  *  In-efficient `add` for `IArray`/`Array` 
  * Impossible to create a specialized `Builder[T, IArray[T]]` for primitive types

I'm inclined to have the `IArrayBuilder[T]` for performance reasons. Otherwise, we might see applications that use the `ArrayBuilder` and then cast it to `IArray` (or use `IArray.unsafeFromArray`)

Comment (sjrd) [reply library/src/scala/collection/mutable/IArrayBuilder.scala:42]: > Impossible to create a specialized Builder[T, IArray[T]] for primitive types

Hum, that's not the case. If you go through `ArrayBuilder.make[T]` followed by a `mapResult`, you will receive the specialized `ArrayBuilder`, which will create a specialized `Array[T]`, and the only thing is that is going to be cast to an `IArray[T]` at the end of the day, which even when not specialized is a no-op. So we would definitely have the specialized builders for `IArray`s.

Now we couldn't have `addAll` specialized for `IArray` and `Array`, that's true. However `addAll` is virtually never used on `Builder`s.

Comment (nicolasstucki) [reply library/src/scala/collection/mutable/IArrayBuilder.scala:42]: The issue was alluding to is that the result of [`mapResult`](https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/mutable/Builder.scala#L81) does not have an overload of `addAll` that receives Arrays.

But if it is almost never used I will remove this.

**Review by sjrd — APPROVED** [review]

**Review by odersky — APPROVED** [review]

File: library/src/scala/IArray.scala [hunk 11]
```diff
@@ -363,6 +424,25 @@ object IArray:
...
+   *  val a: Array[Int] = Array(1, 5)
+   *
+   *  scala> val b = Array.from(Range(1, 5))
+   *  val b: Array[Int] = Array(1, 2, 3, 4)
+   *  }}}
+   *
+   *  @param  it the iterable collection
+   *  @return    an array consisting of elements of the iterable collection
+   */
+  def from[A : ClassTag](it: IterableOnce[A]): Array[A] =
```
Comment (bishabosha) [inline library/src/scala/IArray.scala:440]: why does this return an `Array`?

Comment (sjrd) [reply library/src/scala/IArray.scala:440]: Ouch. Nope.

Comment (nicolasstucki) [reply library/src/scala/IArray.scala:440]: I'll fix it

Comment (nicolasstucki) [reply library/src/scala/IArray.scala:440]: See #11387
